### PR TITLE
[bond] updated version + added bond-over-grpc integration as feature

### LINKED ISSUE #10316: bond - please support building with DBOND_ENABLE_GRPC=TRUE
****

**Is your feature request related to a problem? Please describe.**
Currently editing the portfile manually and speciying DBOND_ENABLE_GRPC=TRUE breaks the build 

```
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: ninja -v
    Working Directory: E:/foundations/vcpkg/buildtrees/bond/x64-windows-rel/vcpkg-parallel-configure
```

From bond documentation it seems that in order to build it requires some 'choco' packages:
activeperl golang ninja yasm
I'm not familiar with cmake so i'm not sure how and if it can be automated but since there's already a port for gRPC i'm guessing this is a viable option.

**Proposed solution**
Support building together with grpc as a package flavor

### PR DESCRIPTION
**Pull request by eitanhs**
**Describe the pull request**

- What does your PR fix?
Fixes #10316

- Which triplets are supported/not supported? Have you updated the CI baseline?
currently tested internally against windows-x64 but should work with other triplets without modifications. Didn't update the CI baseline.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
As much as I could, I'm not a C-Make pro so I'm not sure if and what I could have done better.

### CODE REVIEW

File: ports/bond/portfile.cmake [hunk 1]
```diff
@@ -5,17 +5,17 @@ vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
 vcpkg_from_github(
```
Comment (LilyWangL) [inline ports/bond/portfile.cmake:5]: `include(vcpkg_common_functions)` is deprecated. Please remove it.

File: ports/bond/portfile.cmake [hunk 2]
```diff
@@ -34,6 +34,12 @@ else()
 
 endif()
 
+if ("bond-over-grpc" IN_LIST FEATURES)
```
Comment (LilyWangL) [inline ports/bond/portfile.cmake:37]: The way to handle feature can be replaced by [vcpkg_check_features()](https://github.com/microsoft/vcpkg/blob/a3a6530631df905eb5c0e26d0b20d7d548e0c465/docs/maintainers/vcpkg_check_features.md).

**Review by LilyWangL — CHANGES_REQUESTED** [review]

File: ports/bond/skip-grpc-compilation.patch [hunk 3]
```diff
@@ -0,0 +1,11 @@
+diff --git a/thirdparty/CMakeLists.txt b/thirdparty/CMakeLists.txt
```
Comment (chwarr) [inline ports/bond/skip-grpc-compilation.patch:1]: We would be happy to take this functionality of skipping gRPC compilation upstream were you to contribute it behind a CMake variable.

File: ports/bond/portfile.cmake (hunk 2 above)
Comment (eitanhs) [reply ports/bond/portfile.cmake]: done

File: ports/bond/portfile.cmake (hunk 1 above)
Comment (eitanhs) [reply ports/bond/portfile.cmake:3]: done

**Comment by LilyWangL** [comment]
@eitanhs Now this port build failed on Linux with the following error:
```
CMake Error at stack_build.cmake:21 (message):
  Version 2.3.1, Git revision de2a7b694f07de7e6cf17f8c92338c16286b2878 (8103
  commits) x86_64 hpack-0.33.0

  2020-05-15 11:23:37.473592: [debug] Checking for project config at:
  /mnt/_work/1/s/buildtrees/bond/src/8.2.0-6f74198761/compiler/stack.yaml

  2020-05-15 11:23:37.479173: [debug] Loading project config file stack.yaml

  2020-05-15 11:23:37.499447: [error] Preventing creation of stack root
  '/home/root/.stack/'.  Parent directory '/home/root/' is owned by someone
  else.
```

**Comment by chwarr** [comment]
In the [Bond CI builds themselves](https://github.com/microsoft/bond/blob/d7557f2abac014195ae2c37d05d67a2fec085bdf/tools/ci-scripts/linux/build_cpp-common.zsh#L5), we have to use cmake `-DBOND_STACK_OPTIONS="--allow-different-user" ...`

**Comment by eitanhs** [comment]
> In the [Bond CI builds themselves](https://github.com/microsoft/bond/blob/d7557f2abac014195ae2c37d05d67a2fec085bdf/tools/ci-scripts/linux/build_cpp-common.zsh#L5), we have to use `cmake -DBOND_STACK_OPTIONS="--allow-different-user" ...`

can I turn this flag regardless of the compiling platform? meaning, just add it to the options list in the portfile?

**Comment by chwarr** [comment]
> > In the [Bond CI builds themselves](https://github.com/microsoft/bond/blob/d7557f2abac014195ae2c37d05d67a2fec085bdf/tools/ci-scripts/linux/build_cpp-common.zsh#L5), we have to use `cmake -DBOND_STACK_OPTIONS="--allow-different-user" ...`
> 
> can I turn this flag regardless of the compiling platform? meaning, just add it to the options list in the portfile?

Seems to work fine on Windows. I just tested it with stack version 2.1.3 on Windows 10 (`cd compiler; stack compile --allow-different-user`) and everything built fine.

The [documentation for `--allow-different-user`](https://docs.haskellstack.org/en/stable/yaml_configuration/#allow-different-user) says "POSIX systems only." I guess that means that on non-POSIX systems or just doesn't do anything rather than causing a failure.

**Comment by eitanhs** [comment]
> @eitanhs Now this port build failed on Linux with the following error:
> ...

@LilyWangL, I thought I fixed it by applying the relevant flag @chwarr suggested.
Unfortunatley the build still fails on Azure Pipelines.
I did not see where is the actual error messages in the build? it shows only summary,
Can you post a link please so I can address the issues?

Thanks.

**Comment by LilyWangL** [comment]
@eitanhs The Linux error:
```
CMake Error at stack_build.cmake:21 (message):
  Invalid argument `"--allow-different-user"'

  Auxiliary command not found in path `stack-"--allow-different-user"'

  File does not exist or is not a regular file `"--allow-different-user"'

  Usage: stack [--help] [--version] [--numeric-version]
  [--hpack-numeric-version]

               [--docker*] [--nix*] ([--verbosity VERBOSITY] | [-v|--verbose] |
               [--silent]) [--[no-]time-in-log] [--stack-root STACK-ROOT]
               [--work-dir WORK-DIR] [--[no-]system-ghc] [--[no-]install-ghc]
               [--arch ARCH] [--ghc-variant VARIANT] [--ghc-build BUILD]
               [-j|--jobs JOBS] [--extra-include-dirs DIR] [--extra-lib-dirs DIR]
               [--with-gcc PATH-TO-GCC] [--with-hpack HPACK]
               [--[no-]skip-ghc-check] [--[no-]skip-msys] [--local-bin-path DIR]
               [--setup-info-yaml URL] [--[no-]modify-code-page]
               [--[no-]allow-different-user] [--[no-]dump-logs]
               [--color|--colour WHEN] [--resolver RESOLVER] [--compiler COMPILER]
               [--[no-]terminal] [--stack-colors|--stack-colours STYLES]
               [--terminal-width INT] [--stack-yaml STACK-YAML] [--lock-file ARG]
               COMMAND|FILE
ninja: build stopped: subcommand failed.
```

**Comment by eitanhs** [comment]
Thanks, I'll try to setup a Linux machine so I'll be able to test it locally before pushing...

**Comment by eitanhs** [comment]
I'm unable to re-produce the same error the Azure Piplines get.
I guess I hit this isse: https://github.com/commercialhaskell/stack/issues/2187 (since the pipleines are running inside docker containers(?))
Not sure how to tackle this :(

**Comment by chwarr** [comment]
@LilyWangL, what version of stack is on the vcpkg VMs? I think it's old enough that it doesn't have the `--allow-different-user` switch.

Below is what I see when I pass an invalid switch to stack Version 2.3.1, Git revision de2a7b694f07de7e6cf17f8c92338c16286b2878. Notice that it has `[--[no-]allow-different-user]`. The error that LilyWangL shared doesn't include that in the help text.

The stack [documentation](https://docs.haskellstack.org/en/stable/yaml_configuration/#allow-different-user) says this switch was added in 1.0.1.

We also know that Bond has trouble with [anything below stack 1.5.1](https://github.com/microsoft/bond#dependencies). If the stack version is old, is there a way to get it updated?

```
2020-05-21 9:28:00 AM (2020-05-21 16:28:00Z) :: chwarr@CHWARR-SL2 :: C:\src
  > stack --not-a-valid-switch
Invalid option `--not-a-valid-switch'

Usage: stack.exe [--help] [--version] [--numeric-version]
                 [--hpack-numeric-version] [--docker*] [--nix*]
                 ([--verbosity VERBOSITY] | [-v|--verbose] | [--silent])
                 [--[no-]time-in-log] [--stack-root STACK-ROOT]
                 [--work-dir WORK-DIR] [--[no-]system-ghc] [--[no-]install-ghc]
                 [--arch ARCH] [--ghc-variant VARIANT] [--ghc-build BUILD]
                 [-j|--jobs JOBS] [--extra-include-dirs DIR]
                 [--extra-lib-dirs DIR] [--with-gcc PATH-TO-GCC]
                 [--with-hpack HPACK] [--[no-]skip-ghc-check] [--[no-]skip-msys]
                 [--local-bin-path DIR] [--setup-info-yaml URL]
                 [--[no-]modify-code-page] [--[no-]allow-different-user]
                 [--[no-]dump-logs] [--color|--colour WHEN]
                 [--resolver RESOLVER] [--compiler COMPILER] [--[no-]terminal]
                 [--stack-colors|--stack-colours STYLES] [--terminal-width INT]
                 [--stack-yaml STACK-YAML] [--lock-file ARG] COMMAND|FILE
```

**Comment by BillyONeal** [comment]
https://github.com/microsoft/vcpkg/issues/11569 See also, Bond's generated files are ill formed and don't work on VS2019 version 16.6.

**Comment by chwarr** [comment]
This is fixed in Bond's [master branch](https://github.com/microsoft/bond/blob/master/CHANGELOG.md). I'm publishing a 9.0 release with this fix today.

**Comment by eitanhs** [comment]
> This is fixed in Bond's [master branch](https://github.com/microsoft/bond/blob/master/CHANGELOG.md). I'm publishing a 9.0 release with this fix today.


Thanks @chwarr !
After upgrading to the latest release all tests seems to pass!

**Comment by eitanhs** [comment]
@LilyWangL - I think I addressed your comment. Can you please re-review?

**Review by LilyWangL — APPROVED** [review]

**Review by BillyONeal — APPROVED** [review]

**Pull request merged**